_blog/2024-02-14-writing-async-rust-traits.mdx

```diff
@@ -0,0 +1,318 @@
+```rust
+trait HttpService {
+    async fn fetch(&self, url: Url) -> Body;
+}
+
+// The above would turn into...
+
+trait HttpService {
+    fn fetch(&self, url: Url) -> impl Future<Output = Body>;
+}
+```
+
+A recent change in Rust 1.75 allowed Return Position Impl Trait In Trait (RPITIT - RFCC 1522, 3425).  This is great for us, as it means that we can now use Generic Associated Types (GATs) in a trait! Let’s have a look at the following trait which is just like the previous one, but now has a generic in it:
+
+```rust
+trait HttpService<T> {
+    async fn fetch(&self, url: Url) -> Body<T>;
+}
+
+// The above would turn into...
+
+trait HttpService<T> {
+    type: $<'me, T>: Future<Output = Body<T>>
+    where Self: 'me;
+
+   fn fetch<T>(&self, obj: Url) -> Self::$<'_, T>
+}
+```
+
+Here we have some generic type that implements a generic trait bound, where `Self` is required to be the same lifetime as the inner type.
+
+What about implementing traits for a type, with GATs? You can implement a trait where the type uses them like so:
+
+```rust
+impl<T> Example for HttpService<T> {
+    async fn fetch(&self, obj: Url) -> Body<T> {
+        self.get(obj).await
+    }
+}
+
+// The above would turn into ..
+
+impl<T> Example for HttpService<T> {
+    type $<'me, T> = impl Future<Output = Body<T>> + 'me
+    where Self: 'me;
+
+    fn fetch<T>(&self, obj: Url) -> Self::$<'_, T> {
+        async move { .. }
+    }
+}
+```
+
+You can find more about Generic Associated Types [here.](https://rust-lang.github.io/rfcs/1598-generic_associated_types.html)
+
+### Mixing syntax
+
+Generally speaking, while the above is all well and good you can also mix the syntax to create whatever you need. You don’t necessarily need to 100% implement it one way:
+
+```rust
+trait HttpService: Send {
+    fn fetch(&self, url: Url) -> impl Future<Output = HtmlBody> + Send;
+}
+
+impl HttpService for MyService {
+    async fn fetch(&self, url: Url) -> HtmlBody {
+        self.client.do_fetch(url).await.into_body()
+    }
+}
+```
+
+As you can see here, we add some trait bounds to ensure that the type being implemented always implements `Send` - but then we use async functions inside. 
+
+## Limitations
+
+As you’ve seen, there are a lot of cool different ways you can write async Rust traits! You can do it either the low-level or high-level way, or mix them.
+
+However, `impl Trait` doesn’t allow additional user-defined trait bounds. If you attempt to use RPITIT and then add a second `impl` into the return signature, it won’t compile. This is somewhat of an annoyance that you can fix this by creating a supertrait, by creating a trait that requires two or more traits that you additionally own like so:
+
+```rust
+trait MyFirstTrait {}
+trait MySecondTrait {}
+
+trait MySuperTrait: MyFirstTrait + MySecondTrait {}
+```
+
+Additionally because it’s not object-safe, there is currently no dynamic dispatch - this means that code can become considerably more complex and confusing. 
```

Comment:
For dynamic dispatch we could use the async-trait crate.

Reply:
Ah, I misunderstood, you meant that we can't use dynamic dispatch with RIPTIT because it's  not object-safe. That makes sense!